typing/typecore.ml

```diff
@@ -5817,14 +5898,21 @@ and type_cases
      is to typecheck the guards and the cases, and then to check for some
      warnings that can fire in the presence of guards.
   *)
-  map_half_typed_cases category env ty_arg ty_res loc caselist ~check_if_total
+  map_half_typed_cases ?conts category env ty_arg ty_res loc caselist
+    ~check_if_total
     ~type_body:begin
-      fun { pc_guard; pc_rhs } pat ~ext_env ~ty_expected ~ty_infer
+      fun { pc_guard; pc_rhs } pat ~ext_env ~cont ~ty_expected ~ty_infer
           ~contains_gadt:_ ->
         let guard =
           match pc_guard with
           | None -> None
           | Some scond ->
+            (* It is crucial that the continuation is not used in the
+               `when' expression as the extent of the continuation is
+               yet to be determined. We make the continuation
+               inaccessible by typing the `when' expression using the
+               environment `ext_env' which does not bind the
+               continuation variable. *)
             Some
               (type_expect ext_env scond
                 (mk_expected ~explanation:When_guard Predef.type_bool))
```

Comment:
This comment is erroneous: the `ext_env` environment does bind the continuation variable (and all patterns and module variables). 

And maybe more accidentally, testing with
```ocaml
let strange = match  1 + Effect.perform (Int true) with
  | x -> 1
  | effect Int true, k when Effect.Deep.continue k 0 = 0 -> 3
  | effect Int true, k -> Effect.Deep.continue k 3
```
yields an `Continuation_already_resumed` exception on the third case which is the behavior that I would expect.

Reply:
The comment is valid but the code isn't. See https://github.com/ocaml-multicore/ocaml-multicore/issues/204.

Reply:
Included a fix in https://github.com/ocaml/ocaml/pull/12309/commits/c8e3fc0b98be5d9050bd0fedb4613042d3f13754 with tests.

Reply:
The fix makes sense to me. I was wondering if this would create trouble for mixed cases 
(`exception (Exn k) | effect E, k | A k when ... `) but `when` case are already not supported in the exception-value mixed case.

Reply:
We don't support effect patterns in mixed cases, even without the guard.